### ISSUE TITLE
Scaling Line HoverTool

### ISSUE THREAD

**Issue by AlJohri** [bokeh/bokeh#3642]
Hi, I'm trying to use the HoverTool on the line glyph as per the example shown here: https://github.com/bokeh/bokeh/blob/master/examples/glyphs/glyphs.py

I'm trying to display about 100 lines of timeseries data where each line has maybe 20-30 points. Essentially, I would like to be able to identify the labels for each line and don't care about the value at each point.

I understand the code is currently calculating the x and y value of the nearest point or interpolating a point somewhere on the line but this doesn't seem to scale for more than like 5 or 6 lines.

Is there anyway to disable this functionality and make it do something more similar to the LineLabelTooltip  from mpld3?

http://mpld3.github.io/modules/API.html#mpld3.plugins.LineLabelTooltip

**Comment by bryevdv** [comment]
What do you expect to happen and what is actually  happening? It's not obvious what "does not scale" means, I can imagine a few interpretations. As always, nothing will help diagnosing a problem better than a minimal compete code sample that reproduces the problem.

**Comment by AlJohri** [comment]
Yup @bryevdv sorry for the vagueness. I'll try to work towards a minimal complete example soon. Here is what I can show you for now: 

http://nbviewer.ipython.org/urls/dl.dropboxusercontent.com/u/2623300/bokehtest.ipynb

The page takes about 5-10 seconds to completely load. If you scroll down to "Plot the trajectory of social counts for a random article."

You can see my Bokeh example using the HoverTool with 5 random articles (takes about 2.87 seconds and its rather slow going from line to to line).

And you can see the mpld3 example which is a bit faster (but also it seems the hovering is a bit less accurate).

What I expect to happen: The computation of the tooltip should be fast and not care about the points, (x,y) location, index. It should be able to display the label of 100+ lines rather easily. This isn't quite big data.

What is actually happening: I honestly don't know enough detail to say what's happening. As I stated above, I hypothesize that this has something to do with being point tooltips as opposed to a true line tooltip but I could be mistaken.

I'll work towards a minimal example soon.

**Comment by AlJohri** [comment]
I figured out how to do what I needed with mpld3 so you can close this issue for now. If you're interested in this issue, I'd be happy to help explore/debug it further.

**Comment by bryevdv** [comment]
Line tooltips still have to do point hit testing on every mouse move (to see if the mouse is close to the line or not). I would still be interested in a minimal example.

**Comment by bryevdv** [comment]
Hi @AlJohri I would definitely be interested in more information/a code sample. As a test I created 100 lines with 30 points each and added `HoverTool`. As you can see in the attached gif, the UI response was absolutely instantaneous when moving the mouse. So it's evidently too easy to do something in a bad way, and we need to make it harder to do that, or document something better. 

![ezgif com-crop](https://cloud.githubusercontent.com/assets/1078448/12284876/915410a4-b979-11e5-97b9-19373217a6ea.gif)

I will add that your notebook seems also to be performing as expected. Are the default mode for line hit testing is "nearest" which means the tooltip always snaps to the nearest data point. Are you interpreting that as being "slow"? If you want it to follow the mouse exactly, you want to set `line_policy="interp"` so the hover always follows the mouse exactly.

**Comment by bryevdv** [comment]
Well, I'm on a plane, the mp4 -> gig conversion seems not to have worked. But you can try the code above. The hover response is instantaneous.

**Comment by bryevdv** [comment]
Actually here is a conversion that worked. 

<a href="https://imgflip.com/gif/x9ufo"><img src="https://i.imgflip.com/x9ufo.gif" title="made at imgflip.com"/></a>

I have to assume that sort of instant interactive response is more than sufficient.

**Comment by bryevdv** [comment]
It's also worth considering whether you might want "hline" or "vline" hit testing, instead of point hit testing. Your lines are quite spread out in some places, which means users will have to position the mouse quite precisely to get near enough for the hover to pop up.

**Comment by bryevdv** [comment]
FYI  hit testing performance was specifically a priority for us. We use a spatial index based off of a modified and improved RTree data structure. AFAIK no other library does anything that sophisticated, and our hover tool outperforms any other library I've ever happened to personally try, often dramatically (very visible lag is common with much smaller data than even the example above).

**Comment by AlJohri** [comment]
I'll replicate this tomorrow, thanks for being so thorough!

Sent from Outlook Mobile




On Tue, Jan 12, 2016 at 8:39 PM -0800, "Bryan Van de Ven" <notifications@github.com> wrote:












FYI  hit testing performance was specifically a priority for us. We use a spatial index based off of a modified and improved RTree data structure. AFAIK no other library does anything that sophisticated, and our hover tool outperforms any other library I've ever happened to personally try, often dramatically (very visible lag is common with much smaller data than even the example above).
